src/v/raft/id_allocator_stm.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "raft/consensus.h"
+#include "raft/errc.h"
+#include "raft/logger.h"
+#include "raft/state_machine.h"
+#include "utils/expiring_promise.h"
+#include "utils/mutex.h"
+
+#include <absl/container/flat_hash_map.h>
+
+namespace raft {
+
+// id_allocator is a service to generate cluster-wide unique id (int64)
+// It is based on a state machine running on top of a raft log. For each
+// request (*) it adds "+1" command to the log, reads the log up
+// to this command (identified by a sequence_id), sums all the read
+// commands and returns a result.
+//
+// (*) When we do this procedure for each request it becomes costly
+// in terms of latency so the id_allocator_stm allocates the ids in
+// batches using "+1000" commands (the actual value comes from the
+// configuration see id_allocator_batch_size) and then serves the
+// requests without touching the log until the batch is exhausted.
+// sequence_id is 3-tuple, first component is a node-unique id generated
+// on the app start (we use kvstore to persist the value and guarantee
+// monotonicity / uniqueness), second component is a node id and the
+// last component is persisted in RAM. All the 3 components guarantees
+// uniqueness even if the node is restarted (last component is reset but
+// the first is incremented).
+//
+// Log garbage collection. If we do nothing all the "+1000" commands will
```

Comment:
from the conversation this morning, we'll leave this solution in place, and add the raft machinery for truncating/installing snapshots before building the state machine for transactions.

Reply:
I am looking into this right now and it seems to me that we could simplify this a lot with Raft snapshots.

Reply:
we wouldn't require centralized coordination of log truncations.

Reply:
Good point, I'll  follow up after the idempotent producer feature is done. Mainly because I want to limit the scope and focus on one thing at a time. We haven't used Raft snapshots yet and I don't want to pay the toll of integrating it for the first time in the middle of developing another feature.